File: src/main/java/no/fintlabs/validation/constraints/InstanceValueTypesAreCompatible.java
package no.fintlabs.validation.constraints;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import java.lang.annotation.Retention;

import static java.lang.annotation.RetentionPolicy.RUNTIME;

@Retention(RUNTIME)
@Constraint(validatedBy = InstanceValueTypesAreCompatibleValidator.class)
public @interface InstanceValueTypesAreCompatible {


    String CONFIGURATION_VALUE_TYPE = "configurationValueType";
    String INCOMPATIBLE_INSTANCE_VALUES_KEY_AND_TYPE = "incompatibleInstanceValuesKeyAndType";

    String message() default "contains references to instance values that are incompatible with " +
            "configuration value type={" + CONFIGURATION_VALUE_TYPE + "}" +
            " : {" + INCOMPATIBLE_INSTANCE_VALUES_KEY_AND_TYPE + "}";

    Class<?>[] groups() default {};

    Class<? extends Payload>[] payload() default {};

}

File: src/main/kotlin/no/novari/flyt/configuration/validation/constraints/InstanceValueTypesAreCompatible.kt
package no.novari.flyt.configuration.validation.constraints

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.CLASS)
@Constraint(validatedBy = [InstanceValueTypesAreCompatibleValidator::class])
annotation class InstanceValueTypesAreCompatible(
    val message: String =
        "contains references to instance values that are incompatible with configuration value type={$CONFIGURATION_VALUE_TYPE} : {$INCOMPATIBLE_INSTANCE_VALUES_KEY_AND_TYPE}",
    val groups: Array<KClass<*>> = [],
    val payload: Array<KClass<out Payload>> = [],
) {
    companion object {
        const val CONFIGURATION_VALUE_TYPE = "configurationValueType"
        const val INCOMPATIBLE_INSTANCE_VALUES_KEY_AND_TYPE = "incompatibleInstanceValuesKeyAndType"
    }
}
